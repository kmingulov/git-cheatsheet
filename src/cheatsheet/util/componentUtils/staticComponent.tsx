import React, { Component, ComponentType, ReactElement } from 'react';

export function staticComponent<P extends object>(WrappedComp: ComponentType<P>): ComponentType<P> {
  return class StaticComponent extends Component<P> {
    public static displayName = `staticComponent(${WrappedComp.displayName || WrappedComp.name || ''})`;

    public shouldComponentUpdate(): boolean {
      return false;
    }

    public render(): ReactElement {
      return <WrappedComp { ...this.props as P }/>;
    }
  };
}
