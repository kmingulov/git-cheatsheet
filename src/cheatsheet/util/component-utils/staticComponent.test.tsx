import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React, { Component, ReactElement } from 'react';

import { staticComponent } from './staticComponent';

Enzyme.configure({adapter: new Adapter()});

interface CounterInterface {
  count: number;
}

class Counter extends Component implements CounterInterface {
  public count: number = 0;

  public render(): ReactElement {
    this.count++;

    return (
      <div>{ this.count }</div>
    );
  }
}

const StaticCounter = staticComponent(Counter);

describe(staticComponent, () => {
  it('constructs a valid displayName for components with display names', () => {
    expect(StaticCounter.displayName).toEqual('staticComponent(Counter)');
  });

  it('constructs a valid displayName for components without display names', () => {
    const StaticComponent = staticComponent(() => <div/>);
    expect(StaticComponent.displayName).toEqual('staticComponent()');
  });

  it('renders the wrapped component', () => {
    const component = mount(<StaticCounter/>);
    expect(component).toMatchSnapshot();
  });

  it('doesn\'t re-render the wrapped component', () => {
    const component = mount(<StaticCounter/>);
    expect(component).toMatchSnapshot();

    component.setProps({ prop: 'value' });
    expect(component).toMatchSnapshot();
  });
});
