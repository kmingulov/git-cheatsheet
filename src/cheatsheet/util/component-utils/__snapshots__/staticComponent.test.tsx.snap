// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`staticComponent doesn't re-render the wrapped component 1`] = `
<staticComponent(Counter)>
  <Counter>
    <div>
      1
    </div>
  </Counter>
</staticComponent(Counter)>
`;

exports[`staticComponent doesn't re-render the wrapped component 2`] = `
<staticComponent(Counter)
  prop="value"
>
  <Counter>
    <div>
      1
    </div>
  </Counter>
</staticComponent(Counter)>
`;

exports[`staticComponent renders the wrapped component 1`] = `
<staticComponent(Counter)>
  <Counter>
    <div>
      1
    </div>
  </Counter>
</staticComponent(Counter)>
`;
